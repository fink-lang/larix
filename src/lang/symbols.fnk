{symbol} = import '@fink/prattler/symbols'
{curr_loc} = import '@fink/prattler'

{next_is_new_expr} = import './block/indentation'
{single_expression} = import './block/expr'



infix_led = fn type, op, lbp: fn ctx, left:
  {loc: {start}} = left
  [right, next_ctx] = single_expression:: ctx, lbp
  {loc: {end}} = right

  [{type, op, left, right, loc: {start, end}}, next_ctx]


left_binding = fn lbp: fn ctx:
  match ctx:
    next_is_new_expr:: ?:
      0
    else:
      lbp


non_bining = fn: fn: 0


infix = fn op, type:
  {
    ...symbol:: op

    lbp: left_binding

    led: fn lbp: infix_led:: type, op, lbp + 1
  }


infix_right = fn op, type:
  {
    ...symbol:: op

    lbp: left_binding

    led: fn lbp: infix_led:: type, op, lbp - 1
  }


# TODO: why is this not type, op?
prefix = fn op, type:
  {
    ...symbol:: op

    lbp: non_bining

    nud: fn lbp: fn ctx:
      {start} = curr_loc:: ctx
      [right, next_ctx] = single_expression:: ctx, lbp
      {end} = right.loc

      [{type, op, right, loc: {start, end}}, next_ctx]
  }


literal = fn op, type:
  {
    ...symbol:: op

    lbp: non_bining

    nud: fn: fn ctx:
      loc = curr_loc:: ctx
      # value = curr_value:: ctx
      [{type, value: op, loc}, ctx]
  }


terminator = fn op:
  {
    ...symbol:: op

    lbp: left_binding
  }
