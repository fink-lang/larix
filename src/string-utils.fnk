unindent_text = fn text, ind:
  [first, ...rest] = text.split rx/\n/g

  [...lines] = pipe rest:
    map line:
      line.slice ind

  [first, ...lines].join '\n'
