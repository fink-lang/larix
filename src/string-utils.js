import {curr_loc, next_loc} from '@fink/prattler';


export const unindent_text = (text, ind)=> (
  text
    .split(/\n/g)
    .map((line, idx)=> line.slice(idx > 0 ? ind : 0))
    .join('\n')
);


export const strip_block = (strings, ...parts)=> {
  const [, ...lines] = String.raw({raw: strings}, ...parts).split('\n');
  const ind = lines[0].search(/[^ ]/);
  const str = lines.map((lne)=> lne.slice(ind)).join('\n');
  return str;
};

