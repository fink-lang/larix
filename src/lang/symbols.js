import {symbol as parser_symbol} from '@fink/prattler/symbols';
import {next_loc} from '@fink/prattler';
import {indentation} from './indentation';


export const symbol = (id)=> ({
  ...parser_symbol(id),

  lbp: (lbp)=> (ctx)=> {
    if (next_loc(ctx).start.column <= indentation(ctx)) {
      return 0;
    }
    return lbp;
  }
});
