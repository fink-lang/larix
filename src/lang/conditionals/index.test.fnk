{describe, it, eq, snapshot} = import '../../testing/jest'

{parse_expr} = import '../../'


describe:: 'match ...:', fn:

  it:: 'parses single line: match foo: bar', fn:
    parse_expr(`
      match item:
        test_pattern: result

        # fallback if no match
        else: spam
    `) eq snapshot


  it:: 'parses complex: {foo}: bar', fn:
    parse_expr(`
      match item:
        # nested object pattern
        {foo, spam: {ni}}: bar

        # nested iterables
        [foo, [1, 2]]: bar

        # mixed nested:
        [foo, {bar: 123, spam: [456, 789]}]: ni

        else:
          # fallback for no match
          spam
    `) eq snapshot


