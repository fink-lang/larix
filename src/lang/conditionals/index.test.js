import {other_token} from '@fink/prattler/symbols';

import {parse_expr} from '../../';


import {strip_block} from '../../string-utils';


describe('match ...:', ()=> {
  it('parses single line: match foo: bar', ()=> {
    expect(
      parse_expr(strip_block`
        match item:
          foo: bar
          else: spam
      `)
    ).toEqual({
      type: 'cond',
      op: 'match',
      args: [parse_expr('      item')],
      exprs: [{
        type: 'cond:test:result',
        left: parse_expr('           \n  foo'),
        right: parse_expr('           \n     : bar'),
        loc: {
          start: {pos: 14, line: 2, column: 2},
          end: {pos: 22, line: 2, column: 10}
        }
      }, {
        type: 'cond:test:result',
        left: parse_expr('           \n          \n  else'),
        right: parse_expr('           \n          \n      : spam'),
        loc: {
          start: {pos: 25, line: 3, column: 2},
          end: {pos: 35, line: 3, column: 12}
        }
      }],
      loc: {
        start: {pos: 0, line: 1, column: 0},
        end: {pos: 35, line: 3, column: 12}
      }
    });
  });

  // TODO: multi line match
});

