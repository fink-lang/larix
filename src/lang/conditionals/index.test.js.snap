// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`match ...: parses single line: match foo: bar 1`] = `
cond match (1:0-3:12)
  other (1:6-1:10) item
  :
  cond:test:result (2:2-2:10)
    other (2:2-2:5) foo
    block (2:5-2:10)
      other (2:7-2:10) bar
  cond:test:result (3:2-3:12)
    other (3:2-3:6) else
    block (3:6-3:12)
      other (3:8-3:12) spam
`;
